File: 004/palindrome.go
package main

import "fmt"

func IsPalindrome(number int) bool {
	n := number
	reversed, digit := 0, 0
	for n > 0 {
		digit = n % 10
		reversed = reversed*10 + digit
		n /= 10
	}
	return number == reversed
}

func main() {
	product, max := 0, 0
	for a := 100; a < 1000; a++ {
		for b := a; b < 1000; b++ {
			product = a * b
			if product > max && IsPalindrome(product) {
				max = product
			}
		}
	}
	fmt.Println(max)
}

File: 004/palindrome.rs
fn is_palindrome(number: u64) -> bool {
    let mut n = number;
    let mut reversed = 0;
    while n > 0 {
        let digit = n % 10;
        reversed = reversed * 10 + digit;
        n /= 10;
    }
    number == reversed
}

fn main() {
    let mut largest = 0;
    for a in 100..1000 {
        for b in 100..1000 {
            let product = a * b;
            if product > largest && is_palindrome(product) {
                largest = product;
            }
        }
    }
    println!("{}", largest);
}
